### DISCUSSION TITLE
Lazy Removal from PersistentBag collection (Similar to Add operation)

### DISCUSSION THREAD

**Discussion by gkatoch** [nhibernate/nhibernate-core#2776] [Q&A] [answered]
Hi there,

Is there a way in Bag to remove the items from it in lazy mode similar to add method? The main issue that we have today is when we remove items from the Bag, it initializes all the items in bag and then remove it. This leads to increse in time for bigger datasets as complete data of collection is loaded in memory (Sometimes millions of records).
Similar option is available with Queue, but queue makes the collection item as null instead of removing the item from collection altogether, which again becomes an overhead to handle in source code.
Any leads will be appreciated.
Thanks

**Comment by fredericDelaporte** [comment, marked as answer]
The bag semantic does not allow lazy delete. Use a list, a map or a set instead. Of course if your bag can be mapped as a set, a list or a map, then it is not really a bag.
In usual data-models, most if not all collections really should be mapped as sets.

**Reply by gkatoch** [reply]
Thanks, I can use a Set but Set loads the full collection in-memory before adding any new item since it has to maintain uniqueness. I can override the behavior as my collection will always be of unique  values and then use Set.

**Reply by maca88** [reply]
The set needs to be mapped with `lazy="extra"` and `inverse="true"` in order to avoid loading it when adding a new item.

**Reply by bahusoid** [reply]
> but Set loads the full collection in-memory before adding any new item

Are  you sure? Maybe I'm wrong but if you add transient entity it shouldn't trigger collection loading (there is no need to check existance if NHIbernate knows for sure that entity is new)

**Reply by gkatoch** [reply]
Ok sorry for the confusion in my previous statement. Set loads the full collection in memory **if the Equals method is overwritten (as done in my code)**. The **CanSkipElementExistenceCheck** method checks for override Equals and hence loads it in memory. 
I can tackle that using my implementation and replace the Bag easily with Set. Thanks for the help.
